File: 2025/20250324/소스2.py
#https://leetcode.com/problems/count-days-without-meetings/?envType=daily-question&envId=2025-03-24
#3169. Count Days Without Meetings

class Solution:
    def countDays(self, days: int, meetings: List[List[int]]) -> int:
        #interval 문제
        #merge interval
        #시작 시간으로 정렬
        #겹치면 merge하기

        sorted_meetings = sorted(meetings)
        mystack = [sorted_meetings[0]]
        for meeting in sorted_meetings:
            top_start, top_end = mystack[-1]
            meeting_start, meeting_end = meeting
            if meeting_start <= top_end:
                mystack[-1] = [min(top_start, meeting_start), max(top_end, meeting_end)]
            else:
                mystack.append(meeting)
        
        available_days = days
        for schedule in mystack:
            available_days -= (schedule[1]-schedule[0] + 1 )
        return available_days

File: 2025/20250821/소스2.swift
//https://leetcode.com/problems/count-square-submatrices-with-all-ones/description/?envType=daily-question&envId=2025-08-20
//1277. Count Square Submatrices with All Ones

class Solution {
    func countSquares(_ matrix: [[Int]]) -> Int {
        let n = matrix.count
        let m = matrix[0].count
        var dp = Array(repeating: Array(repeating: 0, count:m), count:n)
        
        for j in 0..<m { 
            dp[0][j] = matrix[0][j]
        }

        for i in 0..<n { 
            dp[i][0] = matrix[i][0]
        }

        for i in 1..<n { 
            for j in 1..<m { 
                if matrix[i][j] == 0 {
                    continue
                }
                if dp[i-1][j] == dp[i][j-1] && dp[i-1][j-1] == dp[i-1][j] {
                    dp[i][j] = dp[i-1][j-1]+1
                } else { 
                    dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1])+1
                }
            }
        }
        
        var sum = 0
        for row in dp { 
            sum += row.reduce(0, {x, y in
                return x + y
            })
        }
        return sum
    }
}